cpp/src/preprocessing/quantize/detail/binary.cuh

```diff
@@ -128,6 +178,77 @@ void transform(raft::resources const& res,
                "the input dataset size (%lu) but is %lu passed",
                dataset_size,
                out_dataset_size);
+  // stats::mean does not support F16
+  RAFT_EXPECTS(!(std::is_same_v<T, half> && params.threshold == binary::set_bit_threshold::mean),
+               "binary::transform does not support threshold == mean for FP16 dataset");
+
+  auto mr = raft::resource::get_workspace_resource(res);
+  auto threshold_vec =
+    raft::make_device_mdarray<T, int64_t>(res, mr, raft::make_extents<int64_t>(0));
+
+  T* threshold_ptr = nullptr;
+
+  if (params.threshold == cuvs::preprocessing::quantize::binary::set_bit_threshold::mean ||
+      params.threshold ==
+        cuvs::preprocessing::quantize::binary::set_bit_threshold::sampling_median) {
+    threshold_vec = raft::make_device_mdarray<T, std::int64_t>(
+      res, mr, raft::make_extents<std::int64_t>(dataset_dim));
+    threshold_ptr = threshold_vec.data_handle();
+  }
+
+  if (params.threshold == cuvs::preprocessing::quantize::binary::set_bit_threshold::mean) {
+    // stats::mean does not support F16
+    if constexpr (!std::is_same_v<T, half>) {
+      raft::stats::mean(
+        res,
+        dataset,
+        raft::make_device_vector_view<T, int64_t>(threshold_vec.data_handle(), dataset_dim),
+        false);
+    }
+  }
+
+  if (params.threshold ==
+      cuvs::preprocessing::quantize::binary::set_bit_threshold::sampling_median) {
+    // Make the number of samples odd so that the median is calculated by only sort and memcpy
+    const size_t num_sampls =
+      std::max(
+        raft::div_rounding_up_safe(static_cast<size_t>(dataset_size * params.sampling_ratio), 2lu) *
+          2lu,
+        2lu) -
+      1;
+    const size_t data_size_per_vector = sizeof(T) * dataset_dim;
+    const uint32_t max_dim_chunk =
+      std::min(std::max(1lu, raft::resource::get_workspace_free_bytes(res) / data_size_per_vector),
+               static_cast<std::size_t>(dataset_dim));
+
+    auto sampled_dataset_chunk = raft::make_device_mdarray<T, int64_t>(
+      res, mr, raft::make_extents<int64_t>(max_dim_chunk, num_sampls));
+    for (uint32_t dim_offset = 0; dim_offset < dataset_dim; dim_offset += max_dim_chunk) {
+      const auto dim_chunk = std::min(max_dim_chunk, dataset_dim - dim_offset);
+
+      sample_and_transpose_dataset<T>(dataset.data_handle() + dim_offset,
```

Comment:
It would be great if we could reuse [raft::matrix::sample_rows()](https://github.com/rapidsai/raft/blob/3ff17a0efc29e96da2db7147549fcbafc4c1f267/cpp/include/raft/matrix/sample_rows.cuh#L39)

Reply:
It is possible to use raft::matrix::sample_rows() here, but would it be okay to keep the current implementation in this PR to reduce memory usage and update it to use the function later, after I add support for strided_matrix as its input in a separate PR for raft?

Reply:
That would make sense.